tools/pyright_tool.py

```diff
@@ -0,0 +1,658 @@
+                print("But we have unused ignores or inconsistencies!")
+                sys.exit(2)
+            else:
+                sys.exit(0)
+        else:
+            print("\nERROR: We have issues!\n")
+            for error in self.real_errors:
+                self.print_human_readable_error(error)
+            print(f"Found {len(self.real_errors)} issues above")
+            if self.unused_ignores or self.inconsistencies:
+                print("And we have unused ignores or inconsistencies!")
+            sys.exit(1)
+
+    def get_and_validate_pyright_config_data(self) -> Dict[str, Any]:
+        """Verify that pyrightconfig exists and has correct data."""
+        if not os.path.isfile(self.pyright_config_file):
+            raise RuntimeError(
+                f"Pyright config file under {self.pyright_config_file} does not exist! "
+                "Tool relies on its existence, please create it."
+            )
+
+        try:
+            config_data = json.loads(open(self.pyright_config_file, "r").read())
+        except json.decoder.JSONDecodeError as err:
+            raise RuntimeError(
+                f"Pyright config under {self.pyright_config_file} does not contain valid JSON! Err: {err}"
+            ) from None
+
+        # enableTypeIgnoreComments MUST be set to False, otherwise the "type: ignore"s
+        # will affect the original pyright result - and we need it to grab all the errors
+        # so we can handle them on our own
+        if (
+            "enableTypeIgnoreComments" not in config_data
+            or config_data["enableTypeIgnoreComments"]
+        ):
+            raise RuntimeError(
+                f"Please set '\"enableTypeIgnoreComments\": true' in {self.pyright_config_file}. "
+                "Otherwise the tool will not work as expected."
+            )
+
+        return config_data
+
+    def get_original_pyright_results(self) -> PyrightResults:
+        """Extract all information from pyright.
+
+        `pyright --outputjson` will return all the results in
+        nice JSON format with `generalDiagnostics` array storing
+        all the errors - schema described in PyrightResults
+        """
+        if self.should_generate_error_file:
+            cmd = f"pyright -p {self.pyright_config_file} --outputjson > {self.error_file}"
+            exit_code = os.system(cmd)
+            # Checking if there was no non-type-checking error when running the above command
+            # From the way `os.system` returns exit codes:
+            # - 0 corresponds to exit code 0 (all fine, no type-checking issues in pyright)
+            # - 256 corresponds to exit code 1 (pyright has found some type-checking issues)
+            # All other exit codes mean something non-type-related got wrong (or pyright was not found)
+            # https://github.com/microsoft/pyright/blob/main/docs/command-line.md#pyright-exit-codes
+            if exit_code not in [0, 256]:
```

Comment:
Nit: when a sequence is not changed during runtime a tuple is usually used `(0, 256)`.